controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java

```diff
@@ -2048,6 +2052,71 @@ private HttpResponse testConfig(ApplicationId id, JobType type) {
+        if (report != null) {
+            Cursor cursor = report.get();
+            // Note: same behaviour for both value '0' and missing value.
+            if (cursor.field("failedAt").asLong() == 0 && cursor.field("completedAt").asLong() == 0) {
+                throw new IllegalArgumentException("Service dump already in progress for " + cursor.field("configId").asString());
+            }
+        }
+        Slime requestPayload;
+        try {
+            requestPayload = SlimeUtils.jsonToSlimeOrThrow(request.getData().readAllBytes());
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Missing or invalid JSON in request content", e);
+        }
+        Cursor requestPayloadCursor = requestPayload.get();
+        String configId = requestPayloadCursor.field("configId").asString();
+        long expiresAt = requestPayloadCursor.field("expiresAt").asLong();
+        if (configId.isEmpty()) {
+            throw new IllegalArgumentException("Missing configId");
+        }
+        Slime dumpRequest = new Slime();
+        Cursor dumpRequestCursor = dumpRequest.setObject();
+        dumpRequestCursor.setLong("createdMillis", controller.clock().millis());
+        dumpRequestCursor.setString("configId", configId);
+        if (expiresAt > 0) {
+            dumpRequestCursor.setLong("expiresAt", expiresAt);
+        }
+        var reportsUpdate = Map.of("serviceDump", new String(uncheck(() -> SlimeUtils.toJsonBytes(dumpRequest))));
+        nodeRepository.updateReports(zone, hostname, reportsUpdate);
+        return new MessageResponse("Request created");
+    }
+
+    private HttpResponse getServiceDump(String tenant, String application, String instance, String environment,
+                                        String region, String hostname, HttpRequest request) {
+        NodeRepository nodeRepository = controller.serviceRegistry().configServer().nodeRepository();
+        ZoneId zone = requireZone(environment, region);
+        Slime report = getReport(nodeRepository, zone, tenant, application, instance, hostname).orElse(null);
+        if (report == null) throw new NotExistsException("No service dump for node " + hostname);
+        return new SlimeJsonResponse(report);
+    }
+
+    private Optional<Slime> getReport(NodeRepository nodeRepository, ZoneId zone, String tenant,
+                                      String application, String instance, String hostname) {
+        Node node;
+        try {
+            node = nodeRepository.getNode(zone, hostname);
+        } catch (IllegalArgumentException e) {
+            throw new NotExistsException(new Hostname(hostname));
+        }
+        ApplicationId app = ApplicationId.from(tenant, application, instance);
+        ApplicationId owner = node.owner().orElseThrow(() -> new IllegalArgumentException("Node has no owner"));
+        if (!app.equals(owner)) {
+            throw new IllegalArgumentException("Node is owned by " + owner.toFullString());
```

Comment:
Probably don't want to leak this info? If the node doesn't exist, doesn't have an owner, or has a different owner should return same generic error.

Reply:
Agree